### ISSUE TITLE
 BB = BB[sorted_ind, :] IndexError: too many indices for array error

### ISSUE THREAD

**Issue by ksai14** [Microsoft/CNTK#2142]
I run A1_GenerateInputROIs.py 
And result ...  

1024 683
Calling selective search..
   Number of rois detected using selective search: 835
   Number of rois on grid added: 93360
   Number of rectangles before filtering  = 94195
   Number of rectangles after filtering  = 3594
   Time [ms]: 4415.981000000001
4 5 negative 5.jpg
Calling selective search..
   Number of rois detected using selective search: 1318
   Number of rois on grid added: 66276
   Number of rectangles before filtering  = 67594
   Number of rectangles after filtering  = 3565
   Time [ms]: 9602.047

Processing image set 'train', image 0 of 25
wrote gt roidb to C:\....\FastRCNN\proc\Grocery_100\cntkFiles\train.cache_gt_roidb.pkl
Only keeping the first 100 ROIs..
wrote ss roidb to C:\....\FastRCNN\proc\Grocery_100\cntkFiles\train.cache_selective_search_roidb.pkl
Number of images in set test = 5
Processing image set 'test', image 0 of 5
wrote gt roidb to C:\....\FastRCNN\proc\Grocery_100\cntkFiles\test.cache_gt_roidb.pkl
Only keeping the first 100 ROIs..
wrote ss roidb to C:\....FastRCNN\proc\Grocery_100\cntkFiles\test.cache_selective_search_roidb.pkl
DONE.

And then

I changed A2_RunWithPyModel.py code 
113 ||  roi_out   = roipooling(conv_out, rois, C.MAX_POOLING, (roi_dim, roi_dim), 0.0625)  to
roi_out   = roipooling(conv_out, rois, (roi_dim, roi_dim))

because  of ''TypeError: roipooling() takes from 3 to 4 positional arguments'' error
then I succeed to run  A2_RunWithPyModel.py 

########################################
Training Fast R-CNN model for 20 epochs.
Training 54603793 parameters in 6 parameter tensors.
Learning rate per sample: 1e-05
Momentum per sample: 0.9048374180359595
Finished Epoch[1 of 20]: [Training] loss = 278.215020 * 25, metric = 51.96% * 25 21.353s (  1.2 samples/s);
Finished Epoch[2 of 20]: [Training] loss = 260.710078 * 25, metric = 23.20% * 25 19.117s (  1.3 samples/s);
.....
.....
Finished Epoch[20 of 20]: [Training] loss = 117.0850 * 25, metric = 23.20% * 25 18.460s (  1.4 samples/s);
Stored trained model at C:\.....\FastRCNN\Output\frcn_py.model
Evaluating Fast R-CNN model for 5 images.

(C:\....\cntk\Examples\Image\Detection\FastRCNN>python 
##############################

And then I run A3_ParseAndEvaluateOutput.py 

but i met another problem 
--
2017-07-20 23:01:46
PARAMETERS: datasetName = Grocery
PARAMETERS: cntk_nrRois = 100
Parsing CNTK output for image set: test
Parsing cntk output file, image 0 of 5
....
Parsing cntk output file, image 4 of 5
test.cache ss roidb loaded from C:\Program Files\Anaconda3\Lib\site-packages\........\FastRCNN\proc\Grocery_100\cntkFiles\test.cache_selective_search_roidb.pkl
   Processing image 0 of 5..
Number of rois before non-maxima surpression: 0
Number of rois  after non-maxima surpression: 0
Evaluating detections
Traceback (most recent call last):
  File "A3_ParseAndEvaluateOutput.py", line 58, in <module>
    evaluate_output()
  File "A3_ParseAndEvaluateOutput.py", line 52, in evaluate_output
    evaluate_net(net, imdb, evalTempDir, None, p.classifier, p.nmsThreshold, boUsePythonImpl=True)
.......
.......
  File "C:\.....\FastRCNN\imdb_data.py", line 257, in _voc _compute PrecisionRecallAp
BB = BB[sorted_ind, :]
IndexError: too many indices for array
####################################

what should I do?

**Comment by kyoro1** [comment]
Hello, @ksai14 .
As I have same trouble with you, I've rollback the code

```py
## current code
roi_out   = roipooling(conv_out, rois, C.MAX_POOLING, (roi_dim, roi_dim), 0.0625)
## What I tried
roi_out   = roipooling(conv_out, rois, (roi_dim, roi_dim))
```
It comes from the history of the repository:
https://github.com/Microsoft/CNTK/commit/9b23b14629bdad54a654640a7cd592e7d164af11

I'm not sure that this is the correct modification, but it might be a kind of walk-through.
Hope it helps you.

**Comment by kyoro1** [comment]
Hello @ksai14 
Tentatively, I tried the omission `BB = BB[sorted_ind, :]` from `imdb_data.py`, i.e. 

```py
## original source:
BB = BB[sorted_ind, :]

## What I changed
#BB = BB[sorted_ind, :]
```

I got a response, but it was just a walk-through. It may not be true correspondence...

**Comment by kyoro1** [comment]
This issue is related to [this one](https://github.com/Microsoft/CNTK/issues/2039)

**Comment by ksai14** [comment]
It seems that the model  does not traine (just guess)

I solved this problem by installing earlierest version(not a Fundamental solution )
https://github.com/Azure/ObjectDetectionUsingCntk#step-4-classifier-training

**Comment by ksai14** [comment]
I just solved this problem change the ./fastRCNN/test.py
line 226 ||  def test_net(net, imdb, output_dir, feature_scale, classifier = 'svm', nmsThreshold = 0.3,
             boUsePythonImpl = False, boThresholdDetections = True , boApplyNms = True):
to

line 226 ||  def test_net(net, imdb, output_dir, feature_scale, classifier = 'svm', nmsThreshold = 0.3,
             boUsePythonImpl = False, boThresholdDetections = False, boApplyNms = True):

but I'm not sure why it is working  and the result is right

**Comment by kyoro1** [comment]
Is your change related to only A3?

**Comment by ksai14** [comment]
....... probably?
and I tried  connect 'selecive selarch.py' to  'Azure/ObjectDetectionUsingCntk'
 because there are confuse  selecive selarch option  in 'Azure/ObjectDetectionUsingCntk'
and i succeed connect each other   (A1_GenerateInputROIs) ==(1_computeRois and 2_cntkGenerateInputs) 
but I met same problem ''BB = BB[sorted_ind, :] IndexError:'' at 5_evaluateResults...

In my opinion ..
the problem is between selective search and model.. ?.,..

**Comment by pkranen** [comment]
Please try with the new CNTK 2.1 release. Closing this one since the user reported that the issue was fixed with the 2.1 release.

**Comment by pavanteja295** [comment]
I'm referring here to **https://github.com/rbgirshick/py-faster-rcnn/** where the same error arises and the reason and the solution for the error is this:

I have seen a lot of posts regarding the error BB = BB[sorted_ind, :] IndexError: too many indices for array which i as well encountered while running the evaluation on a very weakly trained model. I actually found a bug in the file pascal_voc.py which causes this issue.
The all_boxes parameter for the py-faster-rcnn/lib/datasets/pascal_voc.py/evaluate_detection() function in pascal_voc.py is a list of numpy arrays. But in pascal_voc.py/_write_voc_results_file(all_boxes) function in line 251 we check if the

dets == []
but here dets is of type <class 'numpy.ndarray'> and the statement in line 251 evaluates to False leading to consider the empty array being written into the output detection file.

This file with some empty arrays when read in py-faster-rcnn/lib/datasets/eval_voc.py/voc_eval() in line 148 throws and error as BB is just an array.

Solution : In pascal_voc.py/_write_voc_results_file(all_boxes) edit to line 251 should be

dets.tolist() == []

This wont throw the error any more !

Kudos !
